### PR TITLE
Add "filename" string resolver utility

### PR DESCRIPTION
**Pull request by bernardkwok**
Update #997 
Add new utility to perform filename resolving. 
For now:
- Handles resolve of `fileprefix`
- Allows for remapping relative to absolute paths
- Allows for an optional post StringResolver for custom resolving in and able the default resolver.

### CODE REVIEW

File: source/MaterialXFormat/File.cpp [hunk 1]
```diff
@@ -75,7 +77,11 @@ string FilePath::asString(Format format) const
 
     if (format == FormatPosix && isAbsolute())
     {
-        str += "/";
+        // Don't prepend a POSIX separator on a Windows absolute path
+        if (_vec.empty() || !std::isalpha(_vec[0][0]) || (_vec[0][1] != ':'))
```
Comment (niklasharrysson) [inline source/MaterialXFormat/File.cpp:81]: For safety maybe check that _vec[0] is a string of at least 2 characters, before accessing _vec[0][0] and _vec[0][1].

File: source/MaterialXFormat/File.cpp [hunk 2]
```diff
@@ -118,13 +124,28 @@ FilePath FilePath::operator/(const FilePath& rhs) const
...
-    uint32_t result = GetFileAttributes(asString().c_str());
-    return result != INVALID_FILE_ATTRIBUTES;
-#else
-    struct stat sb;
-    return stat(asString().c_str(), &sb) == 0;
-#endif
+//#if defined(_WIN32)
+//    uint32_t result = GetFileAttributes(asString().c_str());
+//    return result != INVALID_FILE_ATTRIBUTES;
+//#else
```
Comment (niklasharrysson) [inline source/MaterialXFormat/File.cpp:130]: I assume this is temporary debug code, so ignoring this.

File: source/MaterialXTest/CMakeLists.txt [hunk 3]
```diff
@@ -144,3 +140,8 @@ set_target_properties(
 target_link_libraries(
     MaterialXTest
     ${CMAKE_DL_LIBS})
+
+install(DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}/../../resources"
+    DESTINATION ${CMAKE_INSTALL_PREFIX} )
+install(DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}/../../resources"
+	DESTINATION ${CMAKE_CURRENT_BINARY_DIR} )
```
Comment (niklasharrysson) [inline source/MaterialXTest/CMakeLists.txt:147]: Is this also debug code, or should we copy the resources into the binary directory?

File: source/MaterialXTest/MaterialXFormat/FormatUtil.cpp [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    mx::DocumentPtr doc1 = mx::createDocument();
+
+    // Set up document
+    mx::NodeGraphPtr nodeGraph = doc1->addNodeGraph();
+    nodeGraph->setFilePrefix(TEST_FILE_PREFIX_STRING.asString() + "\\"); // Note this is required as filepath->string strips out last separator
+    mx::NodePtr image1 = nodeGraph->addNode("image");
+    image1->setInputValue("file", "brass_roughness.jpg", mx::FILENAME_TYPE_STRING);
+    mx::NodePtr image2 = nodeGraph->addNode("image");
+    image2->setInputValue("file", "brass_color.jpg", mx::FILENAME_TYPE_STRING);
+#
```
Comment (niklasharrysson) [inline source/MaterialXTest/MaterialXFormat/FormatUtil.cpp:29]: Is this # supposed to be here?

File: source/MaterialXFormat/Util.h [hunk 5]
```diff
@@ -45,6 +45,14 @@ StringSet loadLibraries(const FilePathVec& libraryFolders,
                         const StringSet& excludeFiles = StringSet(),
                         XmlReadOptions* readOptions = nullptr);
 
+/// Set file name values to resolved values. The resolve invlides apply the Elements resolver, an option relative to absoulte path conversion, and
```
Comment (niklasharrysson) [inline source/MaterialXFormat/Util.h:48]: Some typos: "The resolve invlides apply the Elements resolver, an option relative to absoulte path conversion"

File: source/MaterialXFormat/Util.cpp [hunk 6]
```diff
@@ -137,4 +137,73 @@ StringSet loadLibraries(const FilePathVec& libraryFolders,
...
+{
+    // Set value to name + file prefix.
+    for (ElementPtr elem : doc->traverseTree())
+    {
+        ValueElementPtr valueElem = elem->asA<ValueElement>();
+        if (!valueElem)
+        {
+            continue;
+        }
+        if (valueElem->getType() != FILENAME_TYPE_STRING)
```
Comment (niklasharrysson) [inline source/MaterialXFormat/Util.cpp:150]: This "if" can be merge with the one above.

File: source/MaterialXTest/CMakeLists.txt (hunk 3 above)
Comment (bernardkwok) [reply source/MaterialXTest/CMakeLists.txt:147]: There isn't any change in what get's copied. i.e. We're still copying resources to the install directory and the test directory.
This is just cleanup to remove the add_custom_command() which does the same copy.

File: source/MaterialXFormat/Util.cpp [hunk 7]
```diff
@@ -137,4 +137,73 @@ StringSet loadLibraries(const FilePathVec& libraryFolders,
...
+        // Convert relative to absolute pathing if the file is not alrady found
+        if (!searchPath.isEmpty())
+        {
+            FilePath resolvedValue(resolvedString);
+            if (!resolvedValue.isAbsolute())
+            {
+                for (size_t i = 0; i < searchPath.size(); i++)
+                {
+                    FilePath testPath = searchPath[i] / resolvedValue;
+                    std::cout << "Searching test path: " << testPath.asString() << std::endl;
```
Comment (feldstj) [inline source/MaterialXFormat/Util.cpp:174]: Assuming we don't want these two prints and that this was just for debugging.

File: source/MaterialXTest/MaterialXFormat/FormatUtil.cpp (hunk 4 above)
Comment (bernardkwok) [reply source/MaterialXTest/MaterialXFormat/FormatUtil.cpp:29]: No. Guess a typo.

File: source/MaterialXFormat/File.cpp (hunk 1 above)
Comment (bernardkwok) [reply source/MaterialXFormat/File.cpp:79]: Good idea. I'll fix this.

File: source/MaterialXFormat/Util.cpp (hunk 6 above)
Comment (bernardkwok) [reply source/MaterialXFormat/Util.cpp:150]: Will do.

File: source/MaterialXFormat/File.cpp [hunk 8]
```diff
@@ -41,6 +41,11 @@ const string PATH_LIST_SEPARATOR = ":";
 #endif
 const string MATERIALX_SEARCH_PATH_ENV_VAR = "MATERIALX_SEARCH_PATH";
 
+bool hasWindowsDriveSpecifier(const string& val)
```
Comment (niklasharrysson) [inline source/MaterialXFormat/File.cpp:44]: Nickpick: maybe tag this "inline"

**Review by niklasharrysson — APPROVED** [review]

**Review by feldstj — APPROVED** [review]

**Pull request merged**